src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -0,0 +1,176 @@
+
+  private val indexFile = path + "/index"
+  private val metadataFile = path + "/metadata.json.gz"
+
+  private var elementIdx = 0L
+  private var rootOffset = 0L
+
+  private val rvb = new RegionValueBuilder()
+  private val region = new Region()
+  rvb.set(region)
+
+  private val leafNode = new LeafNodeBuilder(keyType)
+  private val internalNodes = new ArrayBuilder[InternalNodeBuilder]()
+
+  private val trackedOS = new ByteTrackingOutputStream(hConf.unsafeWriter(indexFile))
+  private val codecSpec = CodecSpec.default
+  private val leafEncoder = codecSpec.buildEncoder(leafNode.typ)(trackedOS)
+  private val internalEncoder = codecSpec.buildEncoder(InternalNodeBuilder.typ(keyType))(trackedOS)
+
+  private def calcDepth: Int =
+    math.max(1, (math.log10(elementIdx) / math.log10(branchingFactor)).ceil.toInt) // max necessary for array of length 1 becomes depth = 0
+
+  private def height: Int = math.max(calcDepth - 1, 1) + 1 // ensure always at least one internal level and one leaf level
+
+  private def writeInternalNode(node: InternalNodeBuilder): IndexNodeInfo = {
+    val fileOffset = trackedOS.bytesWritten
+    rootOffset = fileOffset
+
+    val child = node.getChild(0)
+    val firstKey = child.firstKey
+    val firstKeyOffset = child.firstKeyOffset
+    val firstIndex = node.firstIndex
+
+    internalEncoder.writeByte(1)
+
+    val regionOffset = node.write(rvb)
+    internalEncoder.writeRegionValue(region, regionOffset)
+    internalEncoder.flush()
+
+    region.clear()
+    node.clear()
+
+    IndexNodeInfo(fileOffset, firstIndex, firstKey, firstKeyOffset)
+  }
+
+  private def writeLeafNode(idx: Long): IndexNodeInfo = {
+    val fileOffset = trackedOS.bytesWritten
+    val firstKey = leafNode.firstKey
+    val firstOffset = leafNode.firstOffset
+
+    leafEncoder.writeByte(0)
+
+    val regionOffset = leafNode.write(rvb, idx)
+    leafEncoder.writeRegionValue(region, regionOffset)
+    leafEncoder.flush()
+
+    region.clear()
+    leafNode.clear()
+
+    IndexNodeInfo(fileOffset, idx, firstKey, firstOffset)
+  }
+
+  private def write() {
+    def updateInternalNodes(level: Int, info: IndexNodeInfo) {
+      if (level == internalNodes.size)
+        internalNodes += new InternalNodeBuilder(keyType)
+
+      val node = internalNodes(level)
+
+      if (node.size == 0) {
+        node.setFirstIndex(info.firstIndex)
+      }
+
+      node += info
+
+      if (node.size == branchingFactor) {
+        val newNodeInfo = writeInternalNode(node)
+        updateInternalNodes(level + 1, newNodeInfo)
+      }
+    }
+
+    val idx = elementIdx - leafNode.size
+    val newNodeInfo = writeLeafNode(idx)
+    updateInternalNodes(0, newNodeInfo)
+  }
+
+  private def flush() {
+    val nInternalLevels = math.max(calcDepth - 1, 1) // ensure always one internal node
+
+    def flushInternalNodes(level: Int, info: IndexNodeInfo) {
+      if (level != nInternalLevels) {
```

Comment:
I think there is a potential bug here: flushing could cause the tree to grow one level.

I think I'd integrate this code with write: have write take a flush: Boolean argument, and keep going up the tree if the block is full or flush is true and have write skip over empty blocks.  Finally, have write return the offset of the last block it writes, which you can use as the rootOffset in the case flush = true.